### PR TITLE
location export stream to disk

### PR DESCRIPTION
**Pull request by snopoke**
@dimagi/product  This was reverted for now. may not be deployed when creating your weekly mail

### Product Note
The main goal of this was to make the export process use less memory. There may be some slight performance improvements.

User visible changes:
* progress bar for export works better
* 'uncategorized_data' is now shows in a single column instead of one column per key. Multiple values are shown as follows: `key1: value1, key2: value2`
* it is not possible to update uncategorized data during an import
* The process of removing uncategorized data has changed.  Instead of removing the 'uncategorized' data column there is now a 'Delete Uncategorized Data(Y/N)' column which can be filled in to remove uncategorized data.

I've tested this locally with ~600 locations (using `CELERY_ALWAYS_EAGER = False`):
* download locations
* re-upload same file with changes / new locations

### CODE REVIEW

File: corehq/apps/locations/util.py [hunk 1]
```diff
@@ -209,7 +209,7 @@ def get_headers(self):
                 additional_headers.extend(self._prefix_headers('consumption', self.product_codes))
             additional_headers.append('uncategorized_data')
 
-            headers[loc_type.code] = [LOCATION_SHEET_HEADERS.keys()[:-2] + additional_headers]
+            headers[loc_type.code] = [LOCATION_SHEET_HEADERS.values()[:-2] + additional_headers]
```
Comment (emord) [inline corehq/apps/locations/util.py]: I don't think that  slice notation works in python 3 to force the iterator to a list

**Review by emord — APPROVED** [review]
Looks good to me. I think it'll break on py3, but we probably have sufficient tests here to not make that blocking

File: corehq/apps/locations/util.py [hunk 2]
```diff
@@ -241,68 +274,39 @@ def coax_boolean(value):
...
     writer = Excel2007ExportWriter()
-    header_table = [(tab_name, [tab['headers']]) for tab_name, tab in locations]
-    writer.open(header_table=header_table, file=outfile)
-    for tab_name, tab in locations:
-        headers = tab['headers']
-        tab_rows = [[row.get(header, '') for header in headers]
-                    for row in tab['rows']]
-        writer.write([(tab_name, tab_rows)])
+    writer.open(header_table=exporter.get_headers(), file=path)
+    exporter.write_data(writer)
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: Think it's worth sticking this in a try/finally block so the writer is always closed?  It also wouldn't be hard to stick `__enter__` and `__exit__` on `Excel2007ExportWriter` so we can use `with`, but that's probably out of the scope of this PR.  I'll make a note to do that at some point if you don't get to it here.

File: corehq/apps/locations/util.py [hunk 3]
```diff
@@ -123,18 +131,30 @@ def get_location_data_model(domain):
...
     def consumption_dict(self):
         return build_consumption_dict(self.domain)
 
+    def _increment_progress(self):
+        if self._location_count is None:
+            self._location_count = SQLLocation.active_objects.filter(domain=self.domain).count()
+            self._progress_update_chunksize = max(10, self._location_count // 100)
+        self._locations_exported += 1
+        if self._locations_exported % self._progress_update_chunksize == 0:
+            DownloadBase.set_progress(self.async_task, self._locations_exported, self._location_count)
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: This wasn't implemented at all before, right?  I know some GS people who'll be very excited about that.

File: corehq/apps/locations/util.py [hunk 4]
```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
...
+            additional_headers.append(LOCATION_SHEET_HEADERS_OPTIONAL['delete_uncategorized_data'])
+
+            headers[loc_type.code] = [LOCATION_SHEET_HEADERS_BASE.values() + additional_headers]
+
+        return list(headers.items())
+
+    def write_data(self, writer):
+        self._write_type_sheet(writer)
+        for location_type in self.location_types:
+            self._write_locations(writer, location_type)
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: I like this refactor a lot

File: corehq/apps/locations/util.py [hunk 5]
```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
...
+
+    def _write_locations(self, writer, location_type):
+        include_consumption = self.include_consumption_flag and location_type.name not in self.administrative_types
+        query = SQLLocation.active_objects.filter(
+            domain=self.domain,
+            location_type__name=location_type.name
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: I think this should be `location_type=location_type` or `location_type_id=location_type.pk` so it doesn't have to do a join.

File: corehq/apps/locations/util.py [hunk 6]
```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
...
+                if include_consumption:
+                    consumption_data = self.get_consumption(loc)
+                    row.extend([consumption_data[code] for code in self.product_codes])
+
+                row.append(', '.join('{}: {}'.format(*d) for d in uncategorized_data.items()))
+
+                yield row
+                self._increment_progress()
+
+        writer.write([(location_type.code, _row_generator())])
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: Nice, so then `writer` will stream the generator without holding the whole thing in memory?

File: corehq/apps/locations/util.py [hunk 7]
```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
...
+                row = [
+                    loc.location_id,
+                    loc.site_code,
+                    loc.name,
+                    loc.parent.site_code if loc.parent else '',
+                    loc.latitude or '',
+                    loc.longitude or '',
+                    loc.external_id,
+                    '',  # do delete
+                ]
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: nit: I prefer not to use positional orderings for stuff like this, but it's not a big deal.  My ideal is the API from [`csv.DictWriter`](https://docs.python.org/2/library/csv.html#csv.DictWriter).

**Review by esoergel — APPROVED** [review]
Lots of good stuff in here!  Some stylistic notes, but nothing blocking.

File: corehq/apps/locations/util.py (hunk 3 above)
Comment (snopoke) [reply corehq/apps/locations/util.py]: Correct

File: corehq/apps/locations/util.py (hunk 7 above)
Comment (snopoke) [reply corehq/apps/locations/util.py]: Agree but that's a bigger change than I want to make right now

File: corehq/apps/locations/util.py (hunk 6 above)
Comment (snopoke) [reply corehq/apps/locations/util.py]: Correct

File: corehq/apps/locations/util.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/locations/util.py]: I did think about that (making `Excel2007ExportWriter` a context manager), not sure why I didn't as I don't think it's a big lift.

File: corehq/apps/locations/util.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/locations/util.py]: This got removed later when I split `LOCATION_SHEET_HEADERS`

File: corehq/apps/locations/util.py (hunk 5 above)
Comment (snopoke) [reply corehq/apps/locations/util.py]: 407632c

File: corehq/apps/locations/util.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/locations/util.py]: 2d0fecb

File: corehq/apps/locations/bulk_management.py [hunk 8]
```diff
@@ -153,9 +154,10 @@ def __init__(self, name, site_code, location_type, parent_code, location_id,
         #   are changed
         self.needs_save = False
         self.moved_to_root = False
+        self.data_model = None
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/locations/bulk_management.py]: 5032206

**Comment by snopoke** [comment]
@dimagi/product this got discussed over email on the product list but pinging here as well before it gets merged in case you have any feedback.

**Review by esoergel — APPROVED** [review]
Awesome, thanks for the followup!

**Review by emord — APPROVED** [review]

File: corehq/apps/locations/util.py [hunk 9]
```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
...
+                for field in self.data_model.fields:
+                    row.append(model_data.get(field.slug, ''))
+
+                if include_consumption:
+                    consumption_data = self.get_consumption(loc)
+                    row.extend([consumption_data[code] for code in self.product_codes])
+
+                row.append(', '.join('{}: {}'.format(*d) for d in uncategorized_data.items()))
```
Comment (dannyroberts) [inline corehq/apps/locations/util.py]: FYI unicode fix here: https://github.com/dimagi/commcare-hq/pull/19563

**Pull request merged**